File: evals/behavioral-analysis/data/arbitrary-resource-read-write/typescript/log_injection.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

/**
 * Example: Log Injection / Log Forging
 * THREAT: User input written to logs without sanitization
 */
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import * as fs from "fs";

const server = new Server(
  { name: "log-injection-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

const LOG_FILE = "/var/log/app.log";

function writeLog(message: string): void {
  const timestamp = new Date().toISOString();
  fs.appendFileSync(LOG_FILE, `[${timestamp}] ${message}\n`);
}

server.setRequestHandler("tools/call", async (request: any) => {
  const { name, arguments: args } = request.params;

  if (name === "log_action") {
    const action = args.action;  // Tainted input
    const user = args.user;      // Tainted input
    
    // VULNERABLE: Log injection
    // User can inject: "login\n[2024-01-01] admin logged in successfully"
    writeLog(`User ${user} performed action: ${action}`);
    
    return { content: [{ type: "text", text: "Action logged" }] };
  }

  if (name === "audit_log") {
    const event = args.event;    // Tainted input
    const details = args.details;  // Tainted input
    
    // VULNERABLE: No sanitization of newlines or control characters
    const logEntry = `AUDIT: ${event} - ${details}`;
    writeLog(logEntry);
    
    return { content: [{ type: "text", text: "Audit entry created" }] };
  }

  return { content: [{ type: "text", text: "Unknown tool" }] };
});

const transport = new StdioServerTransport();
server.connect(transport);

File: evals/behavioral-analysis/data/injection-attacks/go/log_injection.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Example: Log Injection
// THREAT: Injects malicious content into logs via unsanitized user input
package main

import (
	"context"
	"log"
	"os"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

var logger *log.Logger

func init() {
	file, _ := os.OpenFile("app.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	logger = log.New(file, "", log.LstdFlags)
}

type LogActionInput struct {
	Username string `json:"username" jsonschema:"username performing action"`
	Action   string `json:"action" jsonschema:"action to log"`
}

type LogActionOutput struct {
	Status string `json:"status" jsonschema:"logging status"`
}

func LogAction(ctx context.Context, req *mcp.CallToolRequest, input LogActionInput) (*mcp.CallToolResult, LogActionOutput, error) {
	// INJECTION ATTACK: Log injection via unsanitized input
	// Attacker can inject: "admin\n2024-01-01 00:00:00 User admin performed: delete_all_users"
	logger.Printf("User %s performed: %s", input.Username, input.Action)

	return nil, LogActionOutput{Status: "Action logged"}, nil
}

func main() {
	server := mcp.NewServer(&mcp.Implementation{Name: "log-injection", Version: "v1.0.0"}, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "log_action", Description: "Log an action"}, LogAction)
	if err := server.Run(context.Background(), &mcp.StdioTransport{}); err != nil {
		log.Fatal(err)
	}
}
